src/rust/ide/lib/view/src/debug_scenes/shapes.rs

```diff
@@ -53,6 +54,60 @@ where T:frp::HasOutput<Output=Out>, T:Into<frp::Stream<Out>>, Out:frp::Data {
 }
 
 
+
+// ==================
+// === Mock Types ===
+// ==================
+
+/// Allows the creation of arbitrary unique `Type`s.
+#[derive(Clone,Debug,Default)]
+struct DummyTypeGenerator {
+    type_counter : u32
+}
+
+impl DummyTypeGenerator {
+    fn get_dummy_type(&mut self) -> Type {
+        self.type_counter += 1;
+        Type::from(format!("dummy_type_{}",self.type_counter))
+    }
+}
+
+/// Allows executing a closure for every node in a span tree. Can be used to initialise the debug
+/// scene with type information information about every node.
+fn visit_span_tree_nodes<F:FnMut(&span_tree::Node)>(span_tree:&span_tree::SpanTree, mut f:F) {
+    let mut to_visit = vec![span_tree.root_ref()];
+    loop {
+        match to_visit.pop() {
+            None       => break,
+            Some(node) => {
+                let skip          = node.kind.is_empty();
+                if !skip { f(&node) };
+                to_visit.extend(node.children_iter());
+            }
+        }
+    }
+}
+
+
+/// Allows modifying every node in a span tree through a closure. Can be used to initialise the
+/// span tree with type information for the debug scene.
+fn visit_span_tree_nodes_mut<F:FnMut(&mut span_tree::Node)>(span_tree:&mut span_tree::SpanTree, mut f:F) {
+    visit_node_and_children_mut(&mut span_tree.root, &mut f);
+}
+
+fn visit_node_and_children_mut<F:FnMut(&mut span_tree::Node)>(node: &mut span_tree::Node, f: &mut F) {
+    f(node);
+    node.children.iter_mut().for_each(|child| {
+        visit_node_and_children_mut(&mut child.node, f);
+    })
+}
```

Comment:
These are span_tree utilities - move them to span_tree implementation please. Chat with @farmaazon about the. correct placement AND if there are no ready utils for that as Adam was working a lot on different iterator types.